lib/propolis/src/cpuid.rs

```diff
@@ -0,0 +1,413 @@
+    pub fn remove_all(&mut self, func: u32) {
+        self.map.retain(|ident, _val| ident.0 != func);
+    }
+    pub fn get(&self, ident: Ident) -> Option<&Entry> {
+        self.map.get(&ident)
+    }
+    pub fn get_mut(&mut self, ident: Ident) -> Option<&mut Entry> {
+        self.map.get_mut(&ident)
+    }
+    pub fn is_empty(&self) -> bool {
+        self.map.is_empty()
+    }
+
+    pub fn iter(&self) -> Iter {
+        Iter(self.map.iter())
+    }
+
+    pub fn for_regs(&self, eax: u32, ecx: u32) -> Option<Entry> {
+        if let Some(ent) = self.map.get(&Ident(eax, Some(ecx))) {
+            // Exact match
+            Some(*ent)
+        } else if let Some(ent) = self.map.get(&Ident(eax, None)) {
+            // Function-only match
+            Some(*ent)
+        } else {
+            None
+        }
+    }
+}
+impl From<Set> for Vec<bhyve_api::vcpu_cpuid_entry> {
+    fn from(value: Set) -> Self {
+        let mut out = Vec::with_capacity(value.map.len());
+        out.extend(value.map.iter().map(|(ident, leaf)| {
+            let vce_flags = match ident.1.as_ref() {
+                Some(_) => bhyve_api::VCE_FLAG_MATCH_INDEX,
+                None => 0,
+            };
+            bhyve_api::vcpu_cpuid_entry {
+                vce_function: ident.0,
+                vce_index: ident.1.unwrap_or(0),
+                vce_flags,
+                vce_eax: leaf.eax,
+                vce_ebx: leaf.ebx,
+                vce_ecx: leaf.ecx,
+                vce_edx: leaf.edx,
+                ..Default::default()
+            }
+        }));
+        out
+    }
+}
+
+/// Convert a [vcpu_cpuid_entry](bhyve_api::vcpu_cpuid_entry) into an ([Ident],
+/// [Entry]) tuple, suitable for insertion into a [Set].
+///
+/// This would be implemented as a [From] trait if rust let us.
+pub fn from_raw(value: bhyve_api::vcpu_cpuid_entry) -> (Ident, Entry) {
+    let idx = if value.vce_flags & bhyve_api::VCE_FLAG_MATCH_INDEX != 0 {
+        Some(value.vce_index)
+    } else {
+        None
+    };
+
+    (
+        Ident(value.vce_function, idx),
+        Entry {
+            eax: value.vce_eax,
+            ebx: value.vce_ebx,
+            ecx: value.vce_ecx,
+            edx: value.vce_edx,
+        },
+    )
+}
+
+pub struct Iter<'a>(std::collections::btree_map::Iter<'a, Ident, Entry>);
+impl<'a> Iterator for Iter<'a> {
+    type Item = (&'a Ident, &'a Entry);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
+}
+
+/// Specialize a set of cpuid leafs for provided attributes.
+///
+/// This includes things such as a CPU topology (cores/threads/etc), a given
+/// vCPU ID (APIC, core/thread ID, etc), or other info tidbits.
+#[derive(Default)]
+pub struct Specializer {
+    has_smt: bool,
+    num_vcpu: Option<NonZeroU8>,
+    vcpuid: Option<i32>,
+    vendor_kind: Option<VendorKind>,
+    cpu_topo: BTreeSet<TopoKind>,
+    do_cache_topo: bool,
+}
+impl Specializer {
+    pub fn new() -> Self {
+        Self::default()
```

Comment:
Should `has_smt` default to true or false? (RFD 314's discussion of leaf 0xB proposes enabling SMT in the guest by default.)

Reply:
Those sort of policy decisions are left to the propolis-lib consumer

Reply:
I think that policy makes sense, but as things stand the call to `with_vcpu_count` to set `has_smt` is technically optional, correct? If it's 100% up to the consumer then I'd expect them to have to provide a value for `has_smt` in all cases (probably by passing it as a parameter to `new`).

Reply:
With the cache topology fixing as a small exception, `has_smt` is more or less worthless without `num_vcpu` too, so I think it's fine to group them that way for now?

Honestly I expect the Specializer stuff to evolve as we build more things with it, so I'm not to worried about getting it perfect out of the gate.